File: NetLab-backend/internal/model/radius.go
package model

import "time"

// RADIUS 认证用户状态。
const (
	// RadiusUserStatusEnabled 正常可用。
	RadiusUserStatusEnabled = "enabled"
	// RadiusUserStatusDisabled 停用（认证一律拒绝）。
	RadiusUserStatusDisabled = "disabled"
)

// RADIUS 用户与套餐的关联模式。
const (
	// RadiusLinkModeStatic 仅使用用户本行配置（静态快照）。
	RadiusLinkModeStatic = 0
	// RadiusLinkModeDynamic 本行字段为空/0 时动态继承套餐配置。
	RadiusLinkModeDynamic = 1
)

// NAS 状态。
const (
	// RadiusNasStatusEnabled 允许接入。
	RadiusNasStatusEnabled = "enabled"
	// RadiusNasStatusDisabled 拒绝接入。
	RadiusNasStatusDisabled = "disabled"
)

// RadiusProfile 是 RADIUS 策略套餐，定义一组可复用的限速/并发/绑定策略。
type RadiusProfile struct {
	ID        uint64 `gorm:"primaryKey;autoIncrement" json:"id"`
	Name      string `gorm:"type:varchar(64);not null;uniqueIndex" json:"name"`
	Status    string `gorm:"type:varchar(16);not null;default:'enabled';index" json:"status"`
	AddrPool  string `gorm:"type:varchar(64);not null;default:''" json:"addrPool"`
	ActiveNum int    `gorm:"not null;default:0" json:"activeNum"` // 并发在线数限制，0=不限
	UpRate    int    `gorm:"not null;default:0" json:"upRate"`    // 上行限速 Kbps，0=不限
	DownRate  int    `gorm:"not null;default:0" json:"downRate"`  // 下行限速 Kbps，0=不限
	Domain    string `gorm:"type:varchar(64);not null;default:''" json:"domain"`
	// IPv6PrefixPool 是 NAS 侧 IPv6 前缀池名（Framed-IPv6-Pool，用于 SLAAC）。
	IPv6PrefixPool string `gorm:"type:varchar(64);not null;default:''" json:"ipv6PrefixPool"`
	// DelegatedIpv6PrefixPool 是 DHCPv6-PD 委派前缀池名（RFC 6911），
	// 与 IPv6PrefixPool 用途不同，不可复用同一取值。
	DelegatedIpv6PrefixPool string    `gorm:"type:varchar(64);not null;default:''" json:"delegatedIpv6PrefixPool"`
	BindMac                 bool      `gorm:"not null;default:false" json:"bindMac"`
	BindVlan                bool      `gorm:"not null;default:false" json:"bindVlan"`
	Remark                  string    `gorm:"type:varchar(255);not null;default:''" json:"remark"`
	CreatedAt               time.Time `gorm:"type:timestamptz;not null;default:now()" json:"createdAt"`
	UpdatedAt               time.Time `gorm:"type:timestamptz;not null;default:now()" json:"updatedAt"`

	// 运行时回填字段，不落库。
	UserCount int64 `gorm:"-" json:"userCount"`
}

// TableName 指定 RadiusProfile 的数据库表名。
func (RadiusProfile) TableName() string { return "nb_radius_profiles" }

// RadiusUser 是 RADIUS 认证用户，与后台登录用户（nb_users）完全独立。
// Password 为 AES-256-GCM 密文（运行时解密），因此可支持 PAP/CHAP/MS-CHAPv2/EAP。
type RadiusUser struct {
	ID              uint64  `gorm:"primaryKey;autoIncrement" json:"id"`
	Username        string  `gorm:"type:varchar(64);not null;uniqueIndex" json:"username"`
	Password        string  `gorm:"type:varchar(255);not null;default:''" json:"-"`
	ProfileID       *uint64 `gorm:"index" json:"profileId"`
	ProfileLinkMode int     `gorm:"not null;default:1" json:"profileLinkMode"`
	Realname        string  `gorm:"type:varchar(64);not null;default:''" json:"realname"`
	Email           string  `gorm:"type:varchar(255);not null;default:''" json:"email"`
	Mobile          string  `gorm:"type:varchar(20);not null;default:''" json:"mobile"`
	Address         string  `gorm:"type:varchar(255);not null;default:''" json:"address"`
	// MacAddr 是逗号分隔的绑定 MAC 列表（归一化为小写冒号分隔，见 NormalizeMacList）。
	MacAddr   string `gorm:"type:varchar(1024);not null;default:''" json:"macAddr"`
	Vlanid1   int    `gorm:"not null;default:0" json:"vlanid1"`
	Vlanid2   int    `gorm:"not null;default:0" json:"vlanid2"`
	BindMac   bool   `gorm:"not null;default:false" json:"bindMac"`
	BindVlan  bool   `gorm:"not null;default:false" json:"bindVlan"`
	IpAddr    string `gorm:"type:varchar(64);not null;default:''" json:"ipAddr"`
	IpV6Addr  string `gorm:"type:varchar(128);not null;default:''" json:"ipV6Addr"`
	AddrPool  string `gorm:"type:varchar(64);not null;default:''" json:"addrPool"`
	ActiveNum int    `gorm:"not null;default:0" json:"activeNum"`
	UpRate    int    `gorm:"not null;default:0" json:"upRate"`
	DownRate  int    `gorm:"not null;default:0" json:"downRate"`
	Domain    string `gorm:"type:varchar(64);not null;default:''" json:"domain"`
	// IPv6PrefixPool 是 NAS 侧 IPv6 前缀池名。
	IPv6PrefixPool string `gorm:"type:varchar(64);not null;default:''" json:"ipv6PrefixPool"`
	// DelegatedIpv6Prefix 是静态委派给用户的 IPv6 前缀（RFC 4818）。
	DelegatedIpv6Prefix string `gorm:"type:varchar(128);not null;default:''" json:"delegatedIpv6Prefix"`
	// DelegatedIpv6PrefixPool 是 DHCPv6-PD 委派前缀池名（RFC 6911）。
	DelegatedIpv6PrefixPool string     `gorm:"type:varchar(64);not null;default:''" json:"delegatedIpv6PrefixPool"`
	ExpireTime              time.Time  `gorm:"type:timestamptz;not null;index" json:"expireTime"`
	Status                  string     `gorm:"type:varchar(16);not null;default:'enabled';index" json:"status"`
	Remark                  string     `gorm:"type:varchar(255);not null;default:''" json:"remark"`
	LastOnline              *time.Time `gorm:"type:timestamptz" json:"lastOnline"`
	CreatedAt               time.Time  `gorm:"type:timestamptz;not null;default:now()" json:"createdAt"`
	UpdatedAt               time.Time  `gorm:"type:timestamptz;not null;default:now()" json:"updatedAt"`

	// 运行时回填字段，不落库。
	Profile     *RadiusProfile `gorm:"-" json:"-"`
	OnlineCount int            `gorm:"-" json:"onlineCount"`
}

// TableName 指定 RadiusUser 的数据库表名。
func (RadiusUser) TableName() string { return "nb_radius_users" }

// —— 套餐继承取值器：本行字段优先，为空/0 且动态关联模式时回落到套餐配置。——

// GetUpRate 返回生效的上行限速（Kbps）。
func (u *RadiusUser) GetUpRate() int {
	if u.UpRate > 0 || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.UpRate
	}
	return u.Profile.UpRate
}

// GetDownRate 返回生效的下行限速（Kbps）。
func (u *RadiusUser) GetDownRate() int {
	if u.DownRate > 0 || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.DownRate
	}
	return u.Profile.DownRate
}

// GetActiveNum 返回生效的并发在线数限制。
func (u *RadiusUser) GetActiveNum() int {
	if u.ActiveNum > 0 || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.ActiveNum
	}
	return u.Profile.ActiveNum
}

// GetAddrPool 返回生效的地址池名。
func (u *RadiusUser) GetAddrPool() string {
	if u.AddrPool != "" || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.AddrPool
	}
	return u.Profile.AddrPool
}

// GetDomain 返回生效的域（如华为 domain）。
func (u *RadiusUser) GetDomain() string {
	if u.Domain != "" || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.Domain
	}
	return u.Profile.Domain
}

// GetBindMac 返回是否启用 MAC 绑定校验。
func (u *RadiusUser) GetBindMac() bool {
	if u.BindMac || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.BindMac
	}
	return u.Profile.BindMac
}

// GetBindVlan 返回是否启用 VLAN 绑定校验。
func (u *RadiusUser) GetBindVlan() bool {
	if u.BindVlan || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.BindVlan
	}
	return u.Profile.BindVlan
}

// GetIPv6PrefixPool 返回生效的 IPv6 前缀池名。
func (u *RadiusUser) GetIPv6PrefixPool() string {
	if u.IPv6PrefixPool != "" || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.IPv6PrefixPool
	}
	return u.Profile.IPv6PrefixPool
}

// GetDelegatedIpv6PrefixPool 返回生效的 DHCPv6-PD 委派前缀池名。
func (u *RadiusUser) GetDelegatedIpv6PrefixPool() string {
	if u.DelegatedIpv6PrefixPool != "" || u.ProfileLinkMode == RadiusLinkModeStatic || u.Profile == nil {
		return u.DelegatedIpv6PrefixPool
	}
	return u.Profile.DelegatedIpv6PrefixPool
}

// RadiusNas 是一台 RADIUS 接入设备（NAS/BRAS/AC/交换机）。
// Secret 为 AES-256-GCM 密文存储。
type RadiusNas struct {
	ID         uint64    `gorm:"primaryKey;autoIncrement" json:"id"`
	Name       string    `gorm:"type:varchar(64);not null;default:''" json:"name"`
	Identifier string    `gorm:"type:varchar(128);not null;default:'';index" json:"identifier"` // NAS-Identifier 兜底匹配
	Hostname   string    `gorm:"type:varchar(128);not null;default:''" json:"hostname"`
	Ipaddr     string    `gorm:"type:varchar(64);not null;default:'';index" json:"ipaddr"` // 按源 IP 匹配
	Secret     string    `gorm:"type:varchar(255);not null;default:''" json:"-"`
	CoaPort    int       `gorm:"not null;default:3799" json:"coaPort"`
	Model      string    `gorm:"type:varchar(64);not null;default:''" json:"model"`
	VendorCode string    `gorm:"type:varchar(32);not null;default:''" json:"vendorCode"`
	Status     string    `gorm:"type:varchar(16);not null;default:'enabled'" json:"status"`
	Tags       string    `gorm:"type:varchar(255);not null;default:''" json:"tags"`
	Remark     string    `gorm:"type:varchar(255);not null;default:''" json:"remark"`
	CreatedAt  time.Time `gorm:"type:timestamptz;not null;default:now()" json:"createdAt"`
	UpdatedAt  time.Time `gorm:"type:timestamptz;not null;default:now()" json:"updatedAt"`
}

// TableName 指定 RadiusNas 的数据库表名。
func (RadiusNas) TableName() string { return "nb_radius_nas" }

File: NetLab-frontend/src/types/radius.ts
// RADIUS 认证计费相关的 API 类型定义，与后端 camelCase JSON 对齐。

export interface RadiusUserItem {
  id: number
  username: string
  profileId?: number | null
  profileName: string
  profileLinkMode: number
  realname: string
  email: string
  mobile: string
  address: string
  macAddr: string
  vlanid1: number
  vlanid2: number
  bindMac: boolean
  bindVlan: boolean
  ipAddr: string
  ipV6Addr: string
  addrPool: string
  ipv6PrefixPool: string
  delegatedIpv6Prefix: string
  delegatedIpv6PrefixPool: string
  activeNum: number
  upRate: number
  downRate: number
  domain: string
  expireTime: string
  status: string
  remark: string
  onlineCount: number
  lastOnline?: string | null
  createdAt: string
  updatedAt: string
}

export interface RadiusUserListParams {
  page?: number
  size?: number
  keyword?: string
  status?: string
}

export interface RadiusUserListResult {
  items: RadiusUserItem[]
  total: number
  page: number
  size: number
}

/** 创建/更新认证用户的请求体；password 为空表示不修改（更新场景）。
 * expireTime 可选：创建留空后端默认一年，更新留空保持不变。 */
export interface RadiusUserPayload {
  username: string
  password?: string
  profileId?: number | null
  profileLinkMode?: number
  realname?: string
  email?: string
  mobile?: string
  address?: string
  macAddr?: string
  vlanid1?: number
  vlanid2?: number
  bindMac?: boolean
  bindVlan?: boolean
  ipAddr?: string
  ipV6Addr?: string
  addrPool?: string
  ipv6PrefixPool?: string
  delegatedIpv6Prefix?: string
  delegatedIpv6PrefixPool?: string
  activeNum?: number
  upRate?: number
  downRate?: number
  domain?: string
  expireTime?: string
  status?: string
  remark?: string
}

export interface RadiusProfileItem {
  id: number
  name: string
  addrPool: string
  ipv6PrefixPool: string
  delegatedIpv6PrefixPool: string
  activeNum: number
  upRate: number
  downRate: number
  domain: string
  bindMac: boolean
  bindVlan: boolean
  status: string
  remark: string
  userCount: number
  createdAt: string
  updatedAt: string
}

export interface RadiusProfileListParams {
  page?: number
  size?: number
  keyword?: string
}

export interface RadiusProfileListResult {
  items: RadiusProfileItem[]
  total: number
  page: number
  size: number
}

export interface RadiusProfileOption {
  id: number
  name: string
}

/** 创建/更新套餐的请求体。 */
export interface RadiusProfilePayload {
  name: string
  addrPool?: string
  ipv6PrefixPool?: string
  delegatedIpv6PrefixPool?: string
  activeNum?: number
  upRate?: number
  downRate?: number
  domain?: string
  bindMac?: boolean
  bindVlan?: boolean
  status?: string
  remark?: string
}

export interface RadiusNasItem {
  id: number
  name: string
  identifier: string
  hostname: string
  ipaddr: string
  coaPort: number
  model: string
  vendorCode: string
  status: string
  tags: string
  remark: string
  createdAt: string
  updatedAt: string
}

export interface RadiusNasListParams {
  page?: number
  size?: number
  keyword?: string
}

export interface RadiusNasListResult {
  items: RadiusNasItem[]
  total: number
  page: number
  size: number
}

/** 创建/更新 NAS 设备的请求体；secret 为空表示不修改（更新场景）。 */
export interface RadiusNasPayload {
  name: string
  identifier?: string
  hostname?: string
  ipaddr: string
  secret?: string
  coaPort?: number
  model?: string
  vendorCode?: string
  status?: string
  tags?: string
  remark?: string
}

/** 免认证类型：mac=按 Calling-Station-Id 匹配，ip=按 Framed-IP-Address 匹配（支持 CIDR）。 */
export type RadiusBypassType = 'mac' | 'ip'

export interface RadiusBypassItem {
  id: number
  type: RadiusBypassType
  value: string
  status: string
  remark: string
  createdAt: string
  updatedAt: string
}

export interface RadiusBypassListParams {
  page?: number
  size?: number
  keyword?: string
}

export interface RadiusBypassListResult {
  items: RadiusBypassItem[]
  total: number
  page: number
  size: number
}

/** 创建/更新免认证规则的请求体。 */
export interface RadiusBypassPayload {
  type: RadiusBypassType
  value: string
  status?: string
  remark?: string
}

export interface RadiusSessionItem {
  id: number
  username: string
  nasAddr: string
  nasPaddr: string
  nasId: string
  nasClass: string
  framedIpaddr: string
  framedNetmask: string
  framedIpv6Prefix: string
  framedIpv6Address: string
  delegatedIpv6Prefix: string
  macAddr: string
  nasPort: number
  nasPortId: string
  serviceType: number
  nasPortType: number
  acctSessionId: string
  acctSessionTime: number
  sessionTimeout?: number
  acctInputTotal: number
  acctOutputTotal: number
  acctInputPackets: number
  acctOutputPackets: number
  acctStartTime: string
  lastUpdate: string
}

export interface RadiusSessionListParams {
  page?: number
  size?: number
  username?: string
  nasAddr?: string
  macAddr?: string
}

export interface RadiusSessionListResult {
  items: RadiusSessionItem[]
  total: number
  page: number
  size: number
}

export interface RadiusKickResult {
  success: boolean
  responseCode: string
  target: string
  message: string
}

/** CoA 下发请求体；sessionTimeout 与 filterId 至少提供一项。 */
export interface RadiusCoAPayload {
  sessionTimeout?: number
  filterId?: string
}

/** CoA 下发结果；errorCause/errorCauseText/rttMs 由 NAS 响应决定，可能缺省。 */
export interface RadiusCoAResult {
  success: boolean
  responseCode: string
  target: string
  message: string
  errorCause?: string
  errorCauseText?: string
  rttMs?: number
}

export interface RadiusAccountingItem {
  id: number
  username: string
  nasAddr: string
  nasId: string
  nasClass: string
  framedIpaddr: string
  framedNetmask: string
  framedIpv6Address: string
  framedIpv6Prefix?: string
  delegatedIpv6Prefix: string
  macAddr: string
  nasPort?: number
  nasPortId: string
  serviceType: number
  nasPortType: number
  acctSessionId: string
  acctSessionTime: number
  sessionTimeout: number
  acctInputTotal: number
  acctOutputTotal: number
  acctInputPackets: number
  acctOutputPackets: number
  acctTerminateCause: string
  acctStartTime: string
  acctStopTime?: string | null
  lastUpdate?: string | null
}

export interface RadiusAccountingListParams {
  page?: number
  size?: number
  username?: string
  startTime?: string
  endTime?: string
}

export interface RadiusAccountingListResult {
  items: RadiusAccountingItem[]
  total: number
  page: number
  size: number
}

export interface RadiusAuthLogItem {
  id: number
  username: string
  nasAddr: string
  nasPaddr: string
  macAddr: string
  authType: string
  result: string
  reason: string
  createdAt: string
}

export interface RadiusAuthLogListParams {
  page?: number
  size?: number
  keyword?: string
  result?: string
}

export interface RadiusAuthLogListResult {
  items: RadiusAuthLogItem[]
  total: number
  page: number
  size: number
}

/** 证书类型：server=服务器证书（含私钥），ca=CA 证书。 */
export type RadiusCertType = 'server' | 'ca'

export interface RadiusCertItem {
  id: number
  name: string
  certType: RadiusCertType
  certPem: string
  subject: string
  issuer: string
  serial: string
  fingerprint: string
  notBefore: string
  notAfter: string
  hasKey: boolean
  remark: string
  createdAt: string
  updatedAt: string
}

export interface RadiusCertListParams {
  page?: number
  size?: number
  keyword?: string
  certType?: string
}

export interface RadiusCertListResult {
  items: RadiusCertItem[]
  total: number
  page: number
  size: number
}

/** 创建/更新证书的请求体；更新时 certPem/keyPem 留空表示不替换，certType 不可修改。 */
export interface RadiusCertPayload {
  name: string
  certType?: RadiusCertType
  certPem?: string
  keyPem?: string
  remark?: string
}

/** RADIUS 系统设置（/settings/system）。 */
export interface RadiusSystemSettings {
  enabled: boolean
  bindHost: string
  authPort: number
  acctPort: number
  radsecEnabled: boolean
  radsecPort: number
  radsecCertId: number
  radsecCaCertId: number
}

/** RADIUS 服务行为设置（/settings/server）。 */
export interface RadiusServerSettings {
  messageAuthMode: 'disabled' | 'warn' | 'enforce'
  ignorePassword: boolean
  sessionTimeout: number
  acctInterimInterval: number
  historyDays: number
  rejectDelayMaxRejects: number
  rejectDelayWindowSeconds: number
}

/** EAP 认证方式。 */
export type RadiusEapMethod = 'eap-md5' | 'eap-mschapv2' | 'eap-tls' | 'eap-peap' | 'eap-ttls'

/** EAP 设置（/settings/eap）；enabledHandlers 为逗号分隔的方法名或 "*"。 */
export interface RadiusEapSettings {
  enabled: boolean
  method: RadiusEapMethod
  enabledHandlers: string
  tlsServerCertId: number
  tlsClientCaId: number
  tlsMinVersion: '1.2' | '1.3'
}

/** GET /settings 返回的完整配置。 */
export interface RadiusSettings {
  system: RadiusSystemSettings
  server: RadiusServerSettings
  eap: RadiusEapSettings
}

/** NAS 厂商代码选项（与后端 vendors 包的厂商代码对应）。 */
export const RADIUS_VENDOR_CODES = [
  { value: '', labelKey: 'radius:nas.vendors.default' },
  { value: '2011', labelKey: 'radius:nas.vendors.huawei' },
  { value: '14988', labelKey: 'radius:nas.vendors.mikrotik' },
  { value: '25506', labelKey: 'radius:nas.vendors.h3c' },
  { value: '3902', labelKey: 'radius:nas.vendors.zte' },
  { value: '9', labelKey: 'radius:nas.vendors.cisco' },
  { value: '14823', labelKey: 'radius:nas.vendors.aruba' },
  { value: '6527', labelKey: 'radius:nas.vendors.alcatel' },
  { value: '2636', labelKey: 'radius:nas.vendors.juniper' },
  { value: '10055', labelKey: 'radius:nas.vendors.ikuai' },
  { value: '2352', labelKey: 'radius:nas.vendors.radback' },
] as const
